Add July 2021 minutes (#144)
tc39/tc39-notes: meetings/2021-07/july-15.md

WH: I'm trying to understand your slide. The `fibonacciSequence` function returns an array from which Realm?

CP: It is an array-like, and let us assume that it is from the realm that is executing this code on, which is not the other realm. Let's assume that for the sake of simplicity.

> JHD: I mean, the specific potential issue here is that this will return an array whose `.constructor` is `otherRealmGlobalThis.Array`. That's not `Array`. But this is only an issue If somebody's using instance of on arrays, which we've all been trained not to do for 10 years, 

Reviewer: ```suggestion
JHD: I mean, the specific potential issue here is that this will return an array whose `.constructor` is `otherRealmGlobalThis.Array`. That's not `Array`. But this is only an issue If somebody's using `instanceof` on arrays, which we've all been trained not to do for 10 years.
```

> JHD: Arrays are kind of a unique example here, obviously, because we've all been using a raid on his array for a long time and arrays array. That array is proxies. So it's like one example but we're...

Reviewer: ```suggestion
JHD: Arrays are kind of a unique example here, obviously, because we've all been using `Array.isArray` for a long time and Array.isArray tunnels through proxies. So it's like one example but we're...
```

CP: We haven't explored that. We've focused on the basics knowing that with the basic Foundation of pieces here with the callables, you will able to somehow create more advanced mechanism to Shared bits between the two Realms that could be a membrane could be some other mechanism that you use. You could do it, a Boilerplate to try to share that data without copying the device. But I don't know. I haven't, I don't don't think about that much.

GCL: Okay, that's interesting. I guess - yeah, I guess it'll just have to be thought about separately.

CP:  Again, you will go from throwing an error to allowing that object to be passed around, but you have to explore that at some point.

USA: That's it. Surprisingly, there is nothing else on the queue.

LEO: We've had the topics from Daniel from the last meeting that I still want to give space for them to present. But I at this that we were discussing these constraints I wonder if it's time again that we can ask for For the stage advancement, 

USA: I guess you could ask, I guess you could ask again And in after that we could see if anyone would like to talk about it

> LEO: So Jordan, are you feeling about it? I know you bring the main objection so far.
> 
> JHD: I mean So, I have a few classes of concerns. So one is that this feels like a really weird and Magic API for me like it's very clever solution to the problem for And I hear that, that, it achieves the goals that people who want membranes have because it lets those use a membrane Library, still, but then it gives you better results with it and you know, takes more of the complexity and puts it in the platform. I get back. And hear that amp has a use case for which membranes work, but our Overkill and that this all works in a more granular passion. But the use cases that this proposal Our include those and more. And that's what we went into Stage 2 with to solve all of those problems, regardless of what the original stated motivation was, that was the Assumption, at least that I was under, and I'm assuming that everyone else was under, when it went into stage 2. Then the other classic concerned separate from the like kind of strangeness of this new mechanism. The other concern is that then happens with those use cases. And I've certainly seen some examples like you've shown and heard some arguments that it's easy - if you have direct object access it is easy to get things wrong and that is certainly true. And it is certainly likely that there are alternative approaches than direct object access to another realm that would achieve the goals that I'm concerned about in ways that are harder to get wrong. I'm concerned that if this advances to stage 3 with this new form that has a restricted - like a subset of use cases met, that the sort of implied consensus that we also want to solve those other problems. Even if everyone may not have agreed with that like it would I would say implied is kind of gone. And that means that if I want to try and solve that use case, not only do I have to start from scratch with a new proposal, but I also have to come up with a mechanism that a lot of very experienced and intelligent people have not come up with, and admittedly weren't specifically trying to come up with, but it just kind of it's very strange to me that within stage to the shape of a solution would change so drastically such that it's almost a completely different proposal.

Reviewer: ```suggestion
LEO: So JHD, are you feeling about it? I know you bring the main objection so far.

JHD: I have a few classes of concerns. So one is that this feels like a really weird and magic API for me. It's a very clever solution to the problem and I hear that it achieves the goals that people who want membranes have because it lets them use a membrane library still, but then it gives you better results with it and you know, takes more of the complexity and puts it in the platform. I get that. And I hear that AMP has a use case for which membranes work, but are overkill and that this all works in a more granular way. But this does not solve all of the use cases that this proposal went into Stage 2 with a solution for, regardless of what the original stated motivation was. That was the assumption at least that I was under, and I'm assuming that everyone else was under, when it went into stage 2. Then the other class of concern is separate from the strangeness of this new mechanism: the other concern is what then happens with those use cases. And I've certainly seen some examples like you've shown and heard some arguments that it's easy - if you have direct object access it is easy to get things wrong and that is certainly true. And it is certainly likely that there are alternative approaches than direct object access to another realm that would achieve the goals that I'm concerned about in ways that are harder to get wrong. I'm concerned that if this advances to stage 3 with this new form that has a restricted - like a subset of use cases met, that the sort of implied consensus that we also want to solve those other problems. Even if everyone may not have agreed with that like it would I would say implied is kind of gone. And that means that if I want to try and solve that use case, not only do I have to start from scratch with a new proposal, but I also have to come up with a mechanism that a lot of very experienced and intelligent people have not come up with, and admittedly weren't specifically trying to come up with, but it just kind of it's very strange to me that within stage 2 the shape of a solution would change so drastically such that it's almost a completely different proposal.
```

CP: Can you be more specific about what those use cases you feel that this is not solving? assuming that obviously the one that we just discussed, we're saying doesn't solve that because that's problematic?

> JHD: Your example addresses certainly a big class of the use cases they have which is wanting to reduce the footprint of what I have to cache as first run code, such that I can then write robust code to write code that will run robustly later despite future modification of globalThis and the associated things. I hear the feedback that it's difficult to write that example correctly, but it's entirely possible to write it correctly. And the wrapper that I would write would construct a realm for myself, it would lock it down, maybe even with ses, but it would lock it down in some ways, and then it would I could even find functions and wrap them and like switch the prototypes and so on, there's there's lots of ways that it can be. Be constructed to be so that for my purposes, it will be correct. I'm not claiming that's easy. And the word that most people will fail, 

Reviewer: ```suggestion
JHD: Your example addresses certainly a big class of the use cases they have which is wanting to reduce the footprint of what I have to cache as first run code, such that I can then write robust code to write code that will run robustly later despite future modification of globalThis and the associated things. I hear the feedback that it's difficult to write that example correctly, but it's entirely possible to write it correctly. And the wrapper that I would write would construct a realm for myself, it would lock it down, maybe even with ses, but it would lock it down in some ways, and then it would I could even find functions and wrap them and like switch the prototypes and so on, there's there's lots of ways that it can be. Be constructed to be so that for my purposes, it will be correct. I'm not claiming that's easy.
```

JHD: that's that's not the problem I'm trying to solve, okay. So I think it is impossible to avoid having that problem. Once once you don't run first, all bets off. 

MM: Good. 

CP: Even the Realms can...

> JHD: Yeah you can you can modify because the the way into Realms which is pasta but you're also it's about my concern is about reducing the footprint scope of what I have to cash or hold onto that first run. So that it's all available later and the original Realms gives me one thing I have to grab this currently, I have a library that has 15 million downloads a week that I use for all my shims, that grabs everything and I have to explicitly, add each thing I need like, and then all of the shims have to like yeah it's just it's a it's a convenience and kind of scalability issue whereas Whereas, with the realm approach, I can it and of one and done to some degree. 

Reviewer: ```suggestion
JHD: My concern is about reducing the footprint and scope of what I have to cache or hold onto in that first-run code so that it's all available later, and the original Realms gives me one thing I have to grab. Currently, I have a library that has 15 million downloads a week that I use for all my shims, that grabs everything and I have to explicitly, add each thing I need, and then all of the shims have to load them all. It's a convenience and kind of scalability issue whereas, with the realm approach, I can do it "one and done" to some degree. 
```

> CP: Well, one thing that we want to mention about this particular thing, Jordan is have done some it serves around these as well, in terms of birth, for four months. And if you grab that global object and you use all these do notations to access the things that you want when you eat it, you're going to pay the penalty as well. So that's another thing that at least another reason, why we catch everything because...

Reviewer: i have no idea what this comment about birth, 4 months, and eating it, actually was. @fudco?

> JHD: I am a hundred percent unconcerned with performance until I have a hundred percent insured correctness. correctness. I will happy to cross that bridge when I come to it. 

Reviewer: ```suggestion
JHD: I am 100% unconcerned with performance until I have 100% ensured correctness. I will be happy to cross that bridge when I come to it.
```

CP: We have talked about these multiple time to obviously, many people are doing, we do these, as well as software is extensible, sometimes we talk about having the ability to your simply at any given time in a realm, you can access all the intrinsics. You don't need to create a realm to get the intrinsics of the realm that you're running on. That's better than creating a realm to get and all the other realm transects that are new, you get your own intrinsic in the previous iteration of the realm proposal. We have a getter called intrinsics that returns these object that contains all the intrinsics specified in 262. And those intrinsic could be accessed by the percentage interested in the same name, which wasn't Nice. But we cool explore some of that, but not related to Rome because you wanted to get it from the room that you are on. You don't care about creating a new realm. I believe that's a very elegant solution that we can introduce as that means that anyone can use it.

> JHD: You basically just described my get intrinsics tiebreak. So yeah, I mean, you're right, I don't need all of rounds for that. Use case, I need some, I'd like something for that, use case and Realms would have solved it. I'll be it in a way that requires careful work to do correctly. 

Reviewer: ```suggestion
JHD: You basically just described my `get-intrinsics` implementation. So yeah, I mean, you're right, I don't need all of Realms for that use case. I need something for that use case, and Realms would have solved it - all be it in a way that requires careful work to do correctly. 
```

MM: So would you would you agree that something that gave you all of the original intrinsics from your own realm? When I say original, I don't mean unfortunately Which not I mean of course. 

> JHD: Analyzing at the moment that I've first, like, run my code. 

Reviewer: ```suggestion
JHD: I mean only what is available at the moment that I've first run my code. 
```

JHD: Yes.

MM: great!

CP: And that could be also if you - if you don't know where to put it, I believe we should put it in the realm constructor. 

USA: We have some things on the queue just to give some status. There's been some chat the DC 39, delegates channel, one of the things that I caffeine has mentioned And that I think it's important if Kevin wants to bring it here but, also met you has some that, I think it's interesting for these days. 

MAH: Yeah, I think similar idea, it seems that what Jordan is trying to do is get to the original intrinsics. And there are multiple options. One is a imperative API that you can call that's existing right now. Another one could be to re-explore the standard modules where you would be able to access those original intinsics things by importing them.

JHD: that has the same problem as built-in modules themselves, which is unless I can synchronously access them in scripts, it is not sufficient.

MAH: so, you You want them not only in

> JHD: just like levels, I want to be able to synchronously. Get them everywhere in JavaScript. Yeah, that's what that feature is here about which module system I'm using in nor should the future, 

Reviewer: ```suggestion
JHD: just like globals, I'd want to be able to synchronously get them everywhere in JavaScript.
```

MM: For sure, I would solve it better. 

JHD: Yeah, I would agree with that, the Yes. If so then what are, what is the suggestion based on that?

MM:  We've identified a need with some concrete ideas of possible API approaches for satisfying the need, I think we've got something that qualifies very well for a new stage one proposal. I think we should address that need there and we should realize that since the need can be addressed and that there's nothing about the realm proposal in its current state that impedes that in any way that we should allow the Realms proposal in its current state to proceed to stage 3.

> JHD: so, it cues get accused empty. So I think that seems a reasonable path forward under one - "condition" is the wrong word because that implies leverage but under the one, like in one circumstance, which is that the "direct object access for Realms" path. It does not appear to me that there will ever that will ever be allowed unless the current membership of TC39 completely turns over are so. It's, it would be kind of weird to ask for consensus on stage for stage one on a proposal or a problem that nothing was written up and put on the agenda before. right now. So it'd be nice if that could be granted, But that alone isn't sufficient because the real piece is, I would kind of need to know if any like, in the same way as when realms, Went to stage two. There was, it was, I think clearly telegraphed that there were these potential blocking concerns for the direct object access approach, right? Is there anyone in the room who has any reason to believe that such a proposal modulo, you know, dealing with various sorts of concerns that proposals run across, but that such a proposal would, would not have a path toward stage 4. Does that make sense? What I'm asking, I'm trying to figure out if it's worth me investing my time. 

Reviewer: ```suggestion
JHD: so, the queue is empty. I think that seems a reasonable path forward under one - "condition" is the wrong word because that implies leverage but under one circumstance, which is that the "direct object access for Realms" path does not appear to me that there will ever that will ever be allowed unless the current membership of TC39 completely turns over. It would be kind of weird to ask for consensus on for stage 1 on a proposal or a problem that nothing was written up and put on the agenda before right now. So it'd be nice if that could be granted, but that alone isn't sufficient because the real piece is, I would need to know - in the same way as when realms went to stage 2, it was clearly telegraphed that there were these potential blocking concerns for the direct object access approach, right? Is there anyone in the room who has any reason to believe that such a proposal modulo, you know, dealing with various sorts of concerns that proposals run across, but that such a proposal would, would not have a path toward stage 4. Does that make sense? What I'm asking, I'm trying to figure out if it's worth me investing my time.
```

MM: Write the proposal, you're referring to is get in transits proposal. We Sure, yes. Okay, Good question.

USA: Is it something that you'd like a temperature check on. 

> JHD: I mean, a temperature check seems nice, I think specifically if it would, even though Everyone is always allowed at any time to say, I have an objection or I have a constraint, even though I never mentioned it before, it would feel really bad if I put in time on a proposal like this and then came back in six months, a year, two months, whatever, and was surprised with a sudden constraint, it's happened to me once before it is not a fun thing to do and it's very demotivating. So I really just want to make sure that if there's any possibility that any one, could guess that, that might happen that I could get some hints about it now so that I know whether it's worth putting in my time, this is not really a technical or procedural question, I'm sort of asking you a personal question in somebody. Make sure that I'm not going to about to waste my time.

Reviewer: ```suggestion
JHD: I mean, a temperature check seems nice, I think specifically if it would, even though everyone is always allowed at any time to say, "I have an objection" or "I have a constraint, even though I never mentioned it before", it would feel really bad if I put in time on a proposal like this and then came back in six months, a year, two months, whatever, and was surprised with a sudden constraint. It's happened to me once before, and it is not a fun thing to do and it's very demotivating. So I really just want to make sure that if there's any possibility that any one, could guess that, that might happen that I could get some hints about it now so that I know whether it's worth putting in my time, this is not really a technical or procedural question, I'm sort of asking you a personal question to try to make sure that I'm not about to waste my time.
```

LEO: From my reading as well. I think you re very welcomed at the size of the group to continue a lot of discussion for this and making sure like, we provide feedback and all the work that we need to, to move this forward because I think Mark would be interested. I am assuming Mark who can confirm that right now. 

MM: Yeah. So, absolutely definitely interested in collaborating on this proposal and seeing it move forward. We cannot write a session that stays secure into the future without something like this.

USA: one question from Greg, from the chat is Jordan are you looking for implementer concern?

JHD: Yeah, I mean in particular, since the of the friction that Realms at that has been implemented concerned. So I want to I, particularly for that. Thank you Greg.

> AKI: Real quick. I just want to make sure that you take the time to throw. probably wouldn't want to talk. Technical things are not supposed to be on discourse, maybe try to reach out to people who weren't able to make this kind of very specific time zone Restricted meeting time looking for the set that same feedback of find out ahead of time. If there would be any concerns or if or if there would be any Issues that you should be prepared for before you get started. This could make sure to ask outside of this meeting. There's a lot of people who couldn't be here, who may or may not have opinions. I have no idea. 

Reviewer 2: IDK how I both messed this one up and missed it on first review. Extremely false.

```suggestion
```

JHD: No, in that like in that if somebody has simply mutated the object and I'm asking for the object, I want the object by identity, and so, it wouldn't any sense to me to have like a snapshot of it. It's more akin to like, if someone, if I cache getIntrinsic and then later, somebody says math.pi equals 3. And I get the math.Pi intrinsic. It will give me the value of pi.

SYG: Yes, I think that is unproblematic. What is problematic for memory constraints especially is if we not have to give you back the identity they argue original object with the same identity but the old original actual original object in which case, you know that directly contradicts the need for identity. 

> JHD: If you do array.prototype.slice equals 3. And I with my cache to get intrinsic and I get the intrinsic for a rate of prototype that slice, or sorry. Yeah, Rita project that slice. Wait, hold on. If change the object, let's say the object prototype object.to something else. That's empty. And then I asked for object dot prototype dottostring. I need to get the original toString function. 

Reviewer: ```suggestion
JHD: If change the object, let's say the `Object.prototype` object to something else that's empty, and then I asked for `Object.prototype.toString`, I need to get the "original" `toString` function. 
```

USA:  So now that that's out of the way you can Greg’s  question, which is from stage 3. I guess I really do want to ask for that. Yes, sure. That's why we are here. Can we get a stage three on the realms proposal?

JHD: That's fine with me.

USA: Does anyone want to explicitly support? actually wanted to express support. Thank you.

AKI: Looks like JWK is on the queue to say so.

USA: Jack, do you want to see that load? 

JWK: Yeah, I'm supporting stage 3. We have been waiting for this for too long.

CP:Tell me about it.

### Conclusion/Resolution
- Stage 3
- Bikeshed the name "Realm" for one more meeting.
  - Engines (at least V8, Node, FF) agree **not** to ship unflagged until after bikeshedding

## `getOriginals` for stage 1
Presenter: Jordan Harband (JHD)

USA: Yeah. And next up we have Jordan couldn't do you want to ask for stage 1? Yeah. 

> JHD: So even though I don't yet have a repository, we sort of sketched it out here. I would be happy make one tomorrow after I get some sleep. It's is there any reason that could not be stage one already? I would be more comfortable with sleeping more. 

Reviewer: ```suggestion
JHD: So even though I don't yet have a repository, we sort of sketched it out here. I would be happy to make one tomorrow after I get some sleep. is there any reason that could not be stage 1 already?
```

SYG: So, I don't want to slow down your word, I don't want them walking, so let me hear from you. Do you think that if we wait for an official explain it, I'm proposing everything through grants X3 would that slow down velocity for you?

> JHD: No, it's just a nice signal. I think stage one is a problem statement, right? So the specifics of the solution don't need to be explained or understood just yet but - like I'm happy to write up that explainer, it's more like, if the problem is unclear to someone such that they are not willing to or not comfortable, making it stage one without that explainer, then obviously would need to wait. But there's a possibility, we don't meet tomorrow and so I otherwise I would just say let's revisit it tomorrow. 

Reviewer: ```suggestion
JHD: No, it's just a nice signal. I think stage one is a problem statement, right? So the specifics of the solution don't need to be explained or understood just yet but - like I'm happy to write up that explainer, it's more like, if the problem is unclear to someone such that they are not willing to or not comfortable, making it stage 1 without that explainer, then obviously it would need to wait. But there's a possibility we don't meet tomorrow, otherwise I would just say let's revisit it tomorrow. 
```

GC: I say given the confusion surrounding at least the discussion on this API in Matrix over the last ten minutes. think it would be good to have some. Some some things written down concretely that we can understand before discussing this more, 

CP: yeah, you're on a weekend weekend, trailblaze these and get it next meeting for stage 1 and 2 maybe. I can't even get all the pieces in.

WH: I'm also a bit unclear as to what is going into stage one and I would note that there is a deadline to post documents a number of days in advance of a meeting so we can review them.

> JHD:  Sure. And I mean that and the lack of materials are both valid, Reasons to procedurally, reject stage 1. I just but that doesn't mean you have to reject it but it sounds it sounds like they're not, there's at least a few people that aren't comfortable with stage one.

Reviewer: ```suggestion
JHD:  Sure. And I mean that and the lack of materials are both valid reasons to procedurally reject stage 1 - but that doesn't mean you have to reject it. It sounds like there's at least a few people that aren't comfortable with stage 1 yet.
```

WH: Yeah, but I also want to mention that I do like where you're going with this. 

JWK: So I didn't catch up with the conversation before. What is the use case? Why do we need an API for it? 

> JHD: So that people can write code, that can't be trivially broken by someone later messing with the environment like node core uses this, because if do delete function.prototype.call node core just craps itself and dies immediately. And so there's all over nerdcore. They're trying to slowly patch this problem so that its robust against user modification of the platform and all of my npm packages are written in ways where as long as they run first. In a before user code is mess with anything, then they are robust against user code messing with things. And get intrinsic would be a much more convenient and potentially more performant although that's less of a concern for me way to to write these this robust code.

Reviewer: ```suggestion
JHD: So that people can write code, that can't be trivially broken by someone later messing with the environment like node core uses this, because if I do `delete Function.prototype.call` node core just craps itself and dies immediately. All over node core, they're trying to slowly patch this problem so that its robust against user modification of the platform and all of my npm packages are written in ways where as long as they run first, before user code messes with anything, then they are robust against user code messing with things later. getIntrinsic would be a much more convenient and potentially more performant (although that's less of a concern for me) way to to write this robust code.
```

 JWK:I guess that must be able to mock otherwise it might break visualization.

KG: Right. So the assumption is that you are running first so that you can save off a copy of getOriginals that no one else will thereafter be in a position to modify.

JWK: if they are running with why need you can just grab those by yourself

> JHD: I might Define a function first by the time that functions run a bunch of add code might have run or user extensions in their browser and mess with stuff and the in order for the the body of the function to be robust, I have to have cashed everything that it's going to use, and in to protect methods, I have to call bound to them in a way that they don't rely on.collared. Fine. Being present at runtime. And there's, I can already do this. I have a get intrinsic library that does this, but having it built into the platform, just like the Realms, proposal would make things a lot simpler and harder to get wrong.

Reviewer: ```suggestion
JHD: I might define a function first, but by the time that function runs, a bunch of ad code might have run or user extensions in their browser, and messed with stuff, and in order for the the body of the function to be robust, I have to have cached everything that it's going to use, and in to protect methods, I have to call-bind them in a way that they don't rely on `.call` being present at runtime. I can already do this - I have a `get-intrinsic` library that does this, but having it built into the platform, just like the Realms proposal, would make things a lot simpler and harder to get wrong.
```

JWK: The realms use case seems much larger than getIntrinsics. It seems like a very limited use case. I don't know if it's worth it to do that.

> JHD: Yes, it's a fair point that there is a small number of people that may utilize that API, but I mean the that is a significant percentage of npm, download traffic already uses my get intrinsic Library so like it's the end users will will benefit from that. A lot of end users will benefit from the code even if they don't use it directly while. I think this may be true of realms.

Reviewer: ```suggestion
JHD: Yes, it's a fair point that there is a small number of people that may utilize that API, but I mean the that is a significant percentage of npm's download traffic already uses my `get-intrinsic` library so the end users will benefit from that. A lot of end users will benefit from the code even if they don't use it directly. I think this may also be true of Realms.
```

JWK: One of the use cases I can imagine is that I can get the async function prototype without actually using an async function literal or that might not be able to ship to some old browsers.

JHD: Could you rephrase that question? I'm not sure. I understood 

JWK: some intrinsics require some newer syntax reach, for example, async function. AsyncGeneratorPrototype requires an async generator literal to get. And if we ship it, it breaks on the old browser.

> JHD: True, and my get intrinsic Library, currently handles that by using eval to evaluate. The syntax is such that it won't fail to parse an older browsers, but it will still work into her once, like, eval inside a try catch. It's gross. I would love to not have to ship that and, you know, be able to work in CSP environment, for example, but like the yeah, I mean so so then one day it is getting the native get intrinsic function is at this point already has all of the current syntax. So like that was alright, for a get intrinsic polyfill to solve which way. I 

Reviewer: ```suggestion
JHD: True, and my `get-intrinsic` library currently handles that by using `eval` to evaluate. The syntax is such that it won't fail to parse an older browsers, but it will still work with `eval` inside a try catch. It's gross. I would love to not have to ship that and, you know, be able to work in CSP environments, for example, but like the yeah, I mean so so then one day it is getting the native get intrinsic function is at this point already has all of the current syntax. So like that was alright, for a get intrinsic polyfill to solve which way.
```
i kind of trailed off fixing the end, but i think this still improves the beginning

CZW: does that mean calling the intrinsic get a copy of the get the same intrinsic, chat function. 

JWK: Each room has its different solutions X. So X So, X must be the current realms 

CZW: I mean I mean the article get coloring, the Gathering tricks and made some modifications on it like that adding some property on or either. My not change either in intrinsic. how it works. But it but it can be adding some next say, observable properties only. 

JHD: Yeah. I mean if someone sticks like a foo property, an object that prototype and I the object, a prototype intrinsic, the object, I get will have a foo property. that is totally fine. I'm not trying to get an original snapshot. I'm just trying to be able to get the original functions primarily 

CZW: So people can change the intrinsics when they getting run first and the later on the code cannot get the original intrinsics of a function. They will be seeing the modified one. If they are not...

> JHD: Yeah, you it's impossible and should remain, so to protect against code that runs before you, but it is possible. And should be much easier to protect against code that runs after you.

Reviewer: ```suggestion
JHD: It is impossible and should remain so to protect against code that runs before you - but it is possible and should be much easier to do, to protect against code that runs after you.
```

USA: Do we have any consensus in this? No, Jordan. Do you think you'd like to come back to this next meeting? 

> JHD: Yeah, I think there's no consensus for stage 1 for it, but I haven't heard any obvious blockers for stage 1 except I need to have materials prepared before the deadline. So with that understanding, I will prepare it and consider it not a waste of my time and I will plan to come back at the next meeting with that request. So thank you, everyone. I'm sorry to hijack the realms discussion as much as I did.

Reviewer: ```suggestion

JHD: Yeah, I think there's no consensus for stage 1 for it, but I haven't heard any obvious blockers for stage 1 except I need to have materials prepared before the deadline. So with that understanding, I will prepare it and consider it "not a waste of my time" and I will plan to come back at the next meeting with that request. So thank you, everyone. I'm sorry to hijack the realms discussion as much as I did.
```
